Compiler/src/abstractinterpretation.jl

```diff
@@ -3704,7 +3704,7 @@ function scan_specified_partitions(query::F1, walk_binding_partition::F2,
             @assert lookup_world in partition_validity
             this_rte = query(interp, leaf_binding, leaf_partition)
             if @isdefined(rte)
-                if this_rte === rte
+                if this_rte === rte && !(is_some_guard(kind) ⊻ is_some_guard(binding_kind(leaf_partition))) # Don't merge from no guard to guard
```

Comment:
The same check should apply to backdated const, which needs to be treated uniformly with guard.

Reply:
No, it's not a guard, but there should an is_backdated, which #59735 does add.